internal/api/core/route.go

```diff
@@ -0,0 +1,117 @@
+package core
+
+import (
+	"net/http"
+	"reflect"
+	"regexp"
+
+	"github.com/edulinq/autograder/internal/log"
+)
+
+type Route interface {
+	GetMethod() string
+	GetRegex() *regexp.Regexp
+	GetBasePath() string
+	Handle(response http.ResponseWriter, request *http.Request) error
+}
+
+// Inspired by https://benhoyt.com/writings/go-routing/
+type BaseRoute struct {
+	Method   string
+	BasePath string
+	Regex    *regexp.Regexp
+	Handler  RouteHandler
+}
+
+type APIRoute struct {
+	BaseRoute
+	Description string
+	Request     reflect.Type
+	Response    reflect.Type
+}
+
+func (this *BaseRoute) GetMethod() string {
+	return this.Method
+}
+
+func (this *BaseRoute) GetRegex() *regexp.Regexp {
+	return this.Regex
+}
+
+func (this *BaseRoute) GetBasePath() string {
+	return this.BasePath
+}
+
+func (this *BaseRoute) Handle(response http.ResponseWriter, request *http.Request) error {
+	return this.Handler(response, request)
+}
+
+func (this *APIRoute) GetDescription() string {
+	return this.Description
+}
+
+func NewBaseRoute(method string, basePath string, handler RouteHandler) *BaseRoute {
+	return &BaseRoute{
+		Method:   method,
+		BasePath: basePath,
+		Regex:    regexp.MustCompile("^" + NewEndpoint(basePath) + "$"),
+		Handler:  handler,
+	}
+}
+
+func NewRedirect(method string, basePath string, target string) *BaseRoute {
+	redirectFunc := func(response http.ResponseWriter, request *http.Request) error {
+		return handleRedirect(target, response, request)
+	}
+
+	return &BaseRoute{
+		Method:   method,
+		BasePath: basePath,
+		Regex:    regexp.MustCompile("^" + NewEndpoint(basePath) + "$"),
+		Handler:  redirectFunc,
+	}
+}
+
+func NewAPIRoute(basePath string, apiHandler any, description string) *APIRoute {
```

Comment:
As per a comment last round, descriptions should eventually be parsed from comments and not passed directly.

I also said that you should move the descriptions from the passed-in strings to comments on the handlers,
but you just copied it...
You know how I generally feel about copied code ...
(Bad.)

Reply:
I thought this was a bridge step. I interpreted your original comment that said it can be done on a later PR to mean leave the passed in description and prep by also commenting the description on the handler.

I will resolve this comment after I update the code and you have seen it.